src/libmongoc/src/mongoc/mongoc-structured-log.c

```diff
@@ -0,0 +1,871 @@
+   }
+   return MONGOC_STRUCTURED_LOG_DEFAULT_MAX_DOCUMENT_LENGTH;
+}
+
+bool
+mongoc_structured_log_opts_set_max_levels_from_env (mongoc_structured_log_opts_t *opts)
+{
+   BSON_ASSERT_PARAM (opts);
+
+   bool all_ok = true;
+   mongoc_structured_log_level_t level;
+
+   // Errors are not fatal by default; always reported by return value, and reported the first time only via a log
+   // warning.
+   static int err_count_all_atomic;
+   static int err_count_per_component_atomic[STRUCTURED_LOG_COMPONENT_TABLE_SIZE];
+
+   if (_mongoc_structured_log_get_log_level_from_env ("MONGODB_LOG_ALL", &level, &err_count_all_atomic)) {
+      BSON_ASSERT (mongoc_structured_log_opts_set_max_level_for_all_components (opts, level));
+   } else {
+      all_ok = false;
+   }
+
+   for (int component = 0; component < STRUCTURED_LOG_COMPONENT_TABLE_SIZE; component++) {
+      if (_mongoc_structured_log_get_log_level_from_env (
+             gStructuredLogComponentEnvVars[component], &level, &err_count_per_component_atomic[component])) {
+         BSON_ASSERT (mongoc_structured_log_opts_set_max_level_for_component (
+            opts, (mongoc_structured_log_component_t) component, level));
+      } else {
+         all_ok = false;
+      }
+   }
+
+   return all_ok;
+}
+
+static void
+_mongoc_structured_log_default_handler_open_stream (mongoc_structured_log_default_handler_shared_t *shared,
+                                                    const char *path)
+{
+   // shared->mutex must already be locked
+
+   if (!path || !strcasecmp (path, "stderr")) {
+      // Default or explicit stderr
+      shared->stream = stderr;
+      shared->stream_fclose_on_destroy = false;
+   } else if (!strcasecmp (path, "stdout")) {
+      shared->stream = stdout;
+      shared->stream_fclose_on_destroy = false;
+   } else {
+      FILE *file = fopen (path, "a");
+      if (file) {
+         shared->stream = file;
+         shared->stream_fclose_on_destroy = true;
+      } else {
+         char errmsg_buf[BSON_ERROR_BUFFER_SIZE];
+         const char *errmsg = bson_strerror_r (errno, errmsg_buf, sizeof errmsg_buf);
+         MONGOC_WARNING ("Failed to open log file '%s' with error: '%s'. Logging to stderr instead.", path, errmsg);
+         shared->stream = stderr;
+         shared->stream_fclose_on_destroy = false;
+      }
+   }
+}
+
+static FILE *
+_mongoc_structured_log_default_handler_get_stream (mongoc_structured_log_instance_t *instance)
+{
+   // instance->default_handler_shared->mutex must already be locked
+   {
+      FILE *log_stream = instance->default_handler_shared.stream;
+      if (log_stream) {
+         return log_stream;
+      }
+   }
+   _mongoc_structured_log_default_handler_open_stream (&instance->default_handler_shared,
+                                                       instance->opts.default_handler_path);
+   {
+      FILE *log_stream = instance->default_handler_shared.stream;
+      BSON_ASSERT (log_stream);
```

Comment:
```suggestion
      assert (log_stream); // Use `assert` instead of `BSON_ASSERT` to avoid false-positive error from scan-build.
```

This `BSON_ASSERT` appears to trigger a [scan-build failure](https://spruce.mongodb.com/task/mongo_c_driver_scan_build_matrix_scan_build_macos_14_arm64_clang_patch_9498107105bba48863e3bb06fde2a63f6a4476a1_6760dc6a53363d000736e952_24_12_17_02_05_33/logs?execution=0) on the Evergreen run:

```
/Users/ec2-user/data/mci/036174d0d0a43910160f3503f915ea2b/mongoc/src/libmongoc/src/mongoc/mongoc-structured-log.c:373:7: warning: Stream pointer might be NULL [unix.Stream]
  373 |       BSON_ASSERT (log_stream);
      |       ^~~~~~~~~~~~~~~~~~~~~~~~
/Users/ec2-user/data/mci/036174d0d0a43910160f3503f915ea2b/mongoc/src/libbson/src/bson/bson-macros.h:196:10: note: expanded from macro 'BSON_ASSERT'
  196 |          fprintf (stderr, "%s:%d %s(): precondition failed: %s\n", __FILE__, (int) (__LINE__), BSON_FUNC, #test); \
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

I expect this is a false positive. Suggest using a plain `assert` and adding `#include <assert.h>`.

Using an `assert` appears to pass an Evergreen [patch build](https://spruce.mongodb.com/version/6761790c5603400007c8d903).

Aside: the `evergreen` CLI can be be used to create patch builds outside of GitHub PRs:

```bash
evergreen patch --project=mongo-c-driver --description="try plain assert" \
    --yes --finalize \
    -v all \
    -t "scan-build-macos-14-arm64-clang"
```

Reply:
I've been using the evergreen command line tool this whole time, I just didn't run all variants. Is that the recommended approach? This scan-build tool is new to me, I'm looking into it.

Reply:
Ok, that's already what I've been doing.

Reply:
>  Suggest using a plain assert and adding `#include <assert.h>`

The failure does look to me like a bug in the clang static analyzer. I haven't put in the time to isolate it well enough to file a bug report with the clang team as they recommend. I think the `clang-analyzer-unix.Stream` is confusing the earlier macro parameter FILE* with the stderr FILE* going to fprintf. It's not supposed to throw errors like this for stdio streams.

I think I've got a better solution than avoiding BSON_ASSERT. We can do a refactor that's useful anyway, and it avoids the argument confusion issue by separating the fprintf and the assert condition. This commit I just pushed replaces the failures with an inline noreturn function that contains fprintf and abort. I think it's cleaner, and it offers a convenient symbol to set debug breakpoints on.

https://github.com/mongodb/mongo-c-driver/pull/1795/commits/07f49b2c563fdf215bdadf33a6824235dd3fb792

Reply:
Nice. Refactor LGTM. Resolving.